### PR TITLE
feat: support for oas 3.1

### PR DESCRIPTION
**Pull request by mallachari**
Closes: https://github.com/stoplightio/elements/issues/1116
Requires: https://github.com/stoplightio/json-schema-tree/pull/9 for `Nullable` support

Updates `http-spec` and `types` dependencies
Replaces `openapi-sampler` with `@stoplight/json-schema-sampler`
Updates code to  support JSONSchema7

Note: JSONSchema7 lacks `[k: string]: any;` property type (it was present in JSONSchema4) which restricts it from using custom properties. That's the reason of removing `x-tags` and few others from fixtures (they weren't used anyway).

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -3884,7 +3884,7 @@
   dependencies:
     eslint-config-prettier "^7.1.0"
 
-"@stoplight/http-spec@^3.1.1", "@stoplight/http-spec@^3.2.6":
+"@stoplight/http-spec@^3.1.1":
```
Comment (philsturgeon) [inline yarn.lock:3887]: Can we get rid of this lockfile entry somehow?

File: yarn.lock [hunk 2]
```diff
@@ -17065,7 +17094,7 @@ open@^7.0.2, open@^7.0.3:
     is-docker "^2.0.0"
     is-wsl "^2.1.1"
 
-openapi-sampler@^1.0.0-beta.16, openapi-sampler@^1.0.0-beta.18:
```
Comment (philsturgeon) [inline yarn.lock:17068]: Is this still needing to be here?

**Comment by philsturgeon** [comment]
Here's a test OAS I use to show off some of the new OAS3.1 things, and im looking for exclusiveMinimum support.

You might need to log into the meta workspace for this URL to work, but https://deploy-preview-1155--stoplight-elements-demo.netlify.app/?spec=https://stoplight.io/api/v1/projects/meta/different-openapi-versions/nodes/reference/OAS3-1.yaml?branch=main&deref=optimizedBundle&token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJodHRwczovL3N0b3BsaWdodC5pby9qd3QvY2xhaW1zIjp7IngtaGFzdXJhLWRlZmF1bHQtcm9sZSI6Im5vZGVfZXhwb3J0IiwieC1oYXN1cmEtYWxsb3dlZC1yb2xlcyI6WyJub2RlX2V4cG9ydCJdLCJ4LWhhc3VyYS1icmFuY2gtbm9kZS1pZCI6Ijc3ODAyMDYifSwiaWF0IjoxNjE5NzgyNzcwLCJleHAiOjE2MjAyMTQ3NzAsImlzcyI6InN0b3BsaWdodCIsInN1YiI6Ii91c2Vycy9ib3QifQ.v63zMeZphzvL5LpN836ZagR4B1Pa2C0VY-81jQNxae0#/operations/post-user

The User model is reused on the request and the response, and it defines a minimum age, but we arent showing `>18` anywhere.

```
        age:
          type: number
          description: Hold old is the user at time of signup.
          exclusiveMinimum: 18
          examples:
            - 36
```

File: yarn.lock (hunk 2 above)
Comment (marcelltoth) [reply yarn.lock:17068]: It's a false positive, from `demo`.

File: yarn.lock (hunk 1 above)
Comment (marcelltoth) [reply yarn.lock:3887]: Same as above. I'll soon send a PR trying to fix this but it's just noise, shouldn't bother us for this PR.

File: packages/elements/src/components/Docs/HttpOperation/Body.tsx [hunk 3]
```diff
@@ -47,7 +46,7 @@ export const Body = ({ body: { contents = [], description }, onChange }: BodyPro
 
       {isJSONSchema(schema) && (
         <Box>
-          <JsonSchemaViewer resolveRef={refResolver} schema={schema as JSONSchema4} viewMode="write" hideExamples />
+          <JsonSchemaViewer resolveRef={refResolver} schema={schema} viewMode="write" hideExamples />
```
Comment (marcelltoth) [inline packages/elements/src/components/Docs/HttpOperation/Body.tsx:49]: Finally! 💪

File: packages/elements/src/components/TryIt/FormDataBody.tsx [hunk 4]
```diff
@@ -33,38 +33,36 @@ export const FormDataBody: React.FC<FormDataBodyProps> = ({ specification, value
...
-                  value={value instanceof File ? value : undefined}
-                  onChange={newValue =>
-                    newValue
-                      ? onChangeValues({ ...values, [parameter.name]: newValue })
-                      : onChangeValues(omit(values, parameter.name))
-                  }
-                />
-              );
-            }
+        {mapSchemaPropertiesToParameters(parameters).map(parameter => {
```
Comment (marcelltoth) [inline packages/elements/src/components/TryIt/FormDataBody.tsx:36]: Nice refactoring!

**Review by marcelltoth — APPROVED** [review]
Looks good here!

**Review by philsturgeon — APPROVED** [review]

**Comment by mallachari** [comment]
> The User model is reused on the request and the response, and it defines a minimum age, but we arent showing `>18` anywhere.
> 
> ```
>         age:
>           type: number
>           description: Hold old is the user at time of signup.
>           exclusiveMinimum: 18
>           examples:
>             - 36
> ```

@philsturgeon the issue you spotted occurred because http-spec transformes schema as it was oas2/3 document. For it to be transformed correctly (or rather not transformed as it is already `JSONSchema7`) we need to add `jsonSchemaDialect` property to the document. I've added that change and it looks good now.

File: packages/elements/src/utils/oas/index.ts [hunk 5]
```diff
@@ -26,11 +26,20 @@ const isOas3 = (parsed: unknown): parsed is OpenAPIObject =>
...
+  Number.parseFloat(String((parsed as Partial<{ openapi: unknown }>).openapi)) === 3.1;
+
 const OAS_MODEL_REGEXP = /((definitions|components)\/?(schemas)?)\//;
 export const MODEL_REGEXP = /schemas\//;
 export const OPERATION_REGEXP = /\/operations\/.+|paths\/.+\/(get|post|put|patch|delete|head|options|trace)$/;
 
 export function transformOasToServiceNode(apiDescriptionDocument: unknown) {
+  if (isOas31(apiDescriptionDocument)) {
+    apiDescriptionDocument.jsonSchemaDialect = 'http://json-schema.org/draft-07/schema#';
+    return computeServiceNode(apiDescriptionDocument, oas3SourceMap, transformOas3Service, transformOas3Operation);
```
Comment (marcelltoth) [inline packages/elements/src/utils/oas/index.ts:41]: This is a bit shady to me, as we now are introducing a side-effect to the function (and potentially to `API` as a whole - you can pass the spec object directly). Can I suggest
```suggestion
    return computeServiceNode({...apiDescriptionDocument, jsonSchemaDialect: 'http://json-schema.org/draft-07/schema#'}, oas3SourceMap, transformOas3Service, transformOas3Operation);
```
instead?

Comment (mallachari) [reply packages/elements/src/utils/oas/index.ts:41]: Sure, that works as well

**Pull request merged**